driver-sync/src/test/functional/com/mongodb/client/AbstractSessionsProseTest.java

```diff
@@ -0,0 +1,111 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.MongoClientSettings;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.UpdateOneModel;
+import com.mongodb.client.model.Updates;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandStartedEvent;
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.ClusterFixture.getConnectionString;
+import static com.mongodb.ClusterFixture.getDefaultDatabaseName;
+import static com.mongodb.ClusterFixture.serverVersionAtLeast;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+// Prose tests from https://github.com/mongodb/specifications/tree/master/source/sessions
+public abstract class AbstractSessionsProseTest {
+
+    protected abstract MongoClient getMongoClient(MongoClientSettings settings);
+
+    // Test 13 from https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#test-plan"
+    @Test
+    public void shouldCreateServerSessionOnlyAfterConnectionCheckout() throws InterruptedException {
+        assumeTrue(serverVersionAtLeast(3, 6));
+
+        int minLsidSetSize = Integer.MAX_VALUE;
+
+        for (int i = 0; i < 5; i++) {
+            // given
+            Set<BsonDocument> lsidSet = ConcurrentHashMap.newKeySet();
+            MongoClient client = getMongoClient(MongoClientSettings.builder()
+                    .applyConnectionString(getConnectionString())
+                    .applyToConnectionPoolSettings(builder -> builder.maxSize(1))
+                    .addCommandListener(new CommandListener() {
+                        @Override
+                        public void commandStarted(final CommandStartedEvent event) {
+                            lsidSet.add(event.getCommand().getDocument("lsid"));
+                        }
+                    })
+                    .build());
+            MongoCollection<Document> collection = client.getDatabase(getDefaultDatabaseName()).getCollection(getClass().getName());
+
+            // when executing 8 operations concurrently
+            ExecutorService executor = Executors.newFixedThreadPool(3);
```

Comment:
Is there a reason we don't have 8 threads here? Using 3 with a synchronous driver means that we definitely cannot have all 8 operations executing concurrently.

Reply:
No, that is a bug. Done